<div class="alert alert-info"><big><b>Комментарий ревьюера</b></big><br>Привет! 👋<br><br> Ничего не поменялось, меня всё ещё зовут Дмитрий Кирсанов. <br><br>Комментарии всё ещё обозначены пометкой <b>Комментарий ревьюера</b> и всё ещё содержат описание <br><span class="alert-success">✔️ хорошего, 👍 отличного, 📝 советов</span>,&nbsp;&nbsp;<span class="alert-warning">📌 важного, ⚠️ замечаний, 💡 рекомендаций</span>,&nbsp;&nbsp;<span class="alert-danger">🔧 доработок, 🚫 ошибок или 👀 вопросов.</span><br><br>После того, как замечание будет устранено, я сменю красный цвет на жёлтый, чтобы тебе было удобнее выделять те места, которые требуют доработки. Код, которым я предлагаю выделять твои комментарии:<br><code>&lt;div class=&quot;alert btn-primary&quot;&gt;&lt;b&gt;Комментарий студента:&lt;/b&gt;&lt;br&gt;   Ок   &lt;/div&gt;</code><br>Всё ещё просьба задавать вопросы и описывать изменения в проект.</div>

<div class="alert alert-success"><b>Комментарий ревьюера</b>&nbsp;👍<br>
🎉 Поздравляю с успешным завершением последнего проекта! 🎂
</div>

# Проект по SQL


<div class="alert alert-success"><b>Комментарий ревьюера</b>&nbsp;📝<br>
Наверное ты будешь наполнять своё портфолио проектами. При этом важно будет подобрать такие названия для проектов, которые бы показывали, какие именно задачи ты умеешь решать. В целом курс подходит к концу, и многим важно оформлять портфолио хорошо, это относится не только к данному проекту.
</div>

На анализ была предоставлена база данных крупного сервиса для чтения книг по подписке с информацией о книгах, издательствах, авторах, а так же пользовательских обзорах книг. Необходимо на основе этих данных сформулировать ценностное предложение для нового продукта.

## Цель проекта:

* **Провести анализ базы данных для формирования предложения для нового продукта**

## Задачи проекта:

1. Посчитать, сколько книг вышло после 1 января 2000 года;
2. Для каждой книги посчитать количество обзоров и среднюю оценку;
3. Определить издательство, которое выпустило наибольшее число книг толще 50 страниц;
4. Определить автора с самой высокой средней оценкой книг (50 и более оценок);
5. Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

### Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

## Обзор данных

In [1]:
# импорт библиотек
import pandas as pd
from sqlalchemy import create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
             'pwd': 'Sdf4$2;d-d30pp', # пароль
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432, # порт подключения
             'db': 'data-analyst-final-project-db'} # название базы данных

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                         db_config['pwd'],
                                                         db_config['host'],
                                                         db_config['port'],
                                                         db_config['db']
                                                        )

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

<div class="alert alert-success"><b>Комментарий ревьюера</b>&nbsp;📝<br>
Если задать имена ключам в текстовом шаблоне, то можно гораздо короче и удобнее формировать строки:

    'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)
</div>

Загрузим каждую таблицу и ознакомимся кратко с каждой из них.

In [3]:
# таблица books

query = '''SELECT *
           FROM books
           LIMIT 10;'''

pd.io.sql.read_sql(query, con = engine) 

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   
5        6        257               1st to Die (Women's Murder Club  #1)   
6        7        258               2nd Chance (Women's Murder Club  #2)   
7        8        260              4th of July (Women's Murder Club  #4)   
8        9        563                                   A Beautiful Mind   
9       10        445                                 A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116

<div class="alert alert-success"><b>Комментарий ревьюера</b>&nbsp;<br>
✔️ Хорошо, что ограничиваешь число загружаемых в память локальной машины строк средствами SQL, для крупных таблиц это будет критически важно

💡 При этом также (отдельным запросом) полезно выводить общее число строк таблицы
</div>

Таблица содержит данные о книгах, их названиях и дате публикации, идентификаторах автора и издателя (внешние ключи). Это основая таблица, которая связывается с остальными данными.

In [4]:
# таблица authors

query = '''SELECT *
           FROM authors
           LIMIT 10;'''

pd.io.sql.read_sql(query, con = engine) 

author_id                                author
0          1                            A.S. Byatt
1          2        Aesop/Laura Harris/Laura Gibbs
2          3                       Agatha Christie
3          4                         Alan Brennert
4          5              Alan Moore/David   Lloyd
5          6                            Alan Paton
6          7           Albert Camus/Justin O'Brien
7          8                         Aldous Huxley
8          9    Aldous Huxley/Christopher Hitchens
9         10  Aleksandr Solzhenitsyn/H.T. Willetts

Здесь хранятся имена авторов книг.

In [5]:
# таблица publishers

query = '''SELECT *
           FROM publishers;'''

pd.io.sql.read_sql(query, con = engine) 

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

Здесь хранится информация об издательствах книг.

In [6]:
# таблица ratings

query = '''SELECT *
           FROM ratings
           LIMIT 10;'''

pd.io.sql.read_sql(query, con = engine) 

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4

Информация пользовательских оценках книг.

In [7]:
# таблица  reviews

query = '''SELECT *
           FROM reviews
           LIMIT 10;'''

pd.io.sql.read_sql(query, con = engine) 

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   
5          6        3    lesliegibbs   
6          7        4   valenciaanne   
7          8        4    abbottjames   
8          9        5        npowers   
9         10        5        staylor   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...  
5           Analysis no several cause international.  
6  One there cost another. Say type save. With pe...  
7  Within enough mother. There at system full rec...  
8  Thank now focus realize economy focus fly. Ite...  
9  Game push lot reduce where remember. Including...

Здесь хранятся тексты и идентификаторы обзора книг (не напрямую, но таблица связана с предыдущей таблицей `ratings` через поле `username`).

В целом, база данных выглядит структурированной и компактной, у нас явно выделяется основная таблица с книгами и дополнительные таблицы к ней с авторами, рейтингами, издательствами и обзорами книг. Перейдем непосредственно к решению задач.

<div class="alert alert-success"><b>Комментарий ревьюера</b>&nbsp;📝<br>
Хорошо! Мы познакомились с данными: посмотрели содержимое таблиц и заодно проверили соединение с сервером. Ниже некоторые полезные запросы, которые могут помочь в будущем
</div>

In [8]:
# КОД РЕВЬЮЕРА
# какие ещё таблицы доступны в базе
query = '''

SELECT *
FROM pg_catalog.pg_tables
WHERE schemaname != 'pg_catalog' AND
      schemaname != 'information_schema';

'''

display(pd.io.sql.read_sql(query, con = engine))


# смотрим на типы столбцов в интересующих таблицах
query = '''
SELECT 
    table_name, 
    column_name, 
    data_type, 
    is_nullable
FROM INFORMATION_SCHEMA.COLUMNS 
WHERE table_name IN ('books', 'authors', 'publishers', 'ratings', 'reviews');
'''

display(pd.io.sql.read_sql(query, con = engine))

schemaname           tablename         tableowner tablespace  hasindexes  \
0      public              orders    praktikum_admin       None        True   
1      public              visits    praktikum_admin       None        True   
2      public  advertisment_costs    praktikum_admin       None        True   
3      public                trip  praktikum_student       None       False   
4      public              C:oks  praktikum_student       None       False   
5      public                boks  praktikum_student       None       False   
6      public   finished_lessons1  praktikum_student       None        True   
7      public   finished_lessons2  praktikum_student       None        True   
8      public          publishers    praktikum_admin       None        True   
9      public             authors    praktikum_admin       None        True   
10     public             reviews    praktikum_admin       None        True   
11     public             ratings    praktikum_admin       None        True   
12     public    finished_lessons  praktikum_student       None        True   
13     public               books    praktikum_admin       None        True   

    hasrules  hastriggers  rowsecurity  
0      False        False        False  
1      False        False        False  
2      False        False        False  
3      False        False        False  
4      False        False        False  
5      False        False        False  
6      False        False        False  
7      False        False        False  
8      False         True        False  
9      False         True        False  
10     False         True        False  
11     False         True        False  
12     False        False        False  
13     False         True        False

table_name       column_name data_type is_nullable
0   publishers      publisher_id   integer          NO
1   publishers         publisher      text         YES
2      authors         author_id   integer          NO
3      authors            author      text         YES
4      reviews         review_id   integer          NO
5      reviews           book_id   integer         YES
6      reviews          username      text         YES
7      reviews              text      text         YES
8      ratings         rating_id   integer          NO
9      ratings           book_id   integer         YES
10     ratings          username      text         YES
11     ratings            rating   integer         YES
12       books           book_id   integer          NO
13       books         author_id   integer         YES
14       books             title      text         YES
15       books         num_pages   integer         YES
16       books  publication_date      date         YES
17       books      publisher_id   integer         YES

## SQL-запросы

### Задача №1

Для начала посчитаем, сколько книг вышло после 1 января 2000 года;

In [9]:
# самый простой запрос, считаем число книг, в фильтре определяем дату позднее 1 января 2000 года

query = """
        SELECT COUNT(books) as books_amount
        FROM books
        WHERE CAST (publication_date AS date) > '2000-01-01';
        """

pd.io.sql.read_sql(query, con = engine) 

books_amount
0           819

После 1 января 2000 было издано 819 книг.

<div class="alert alert-success"><b>Комментарий ревьюера</b>&nbsp;<br>
✔️ Тут всё верно. Сработал бы и такой запрос:

    publication_date >= '2000-01-01'

Если следовать вопросу буквально, то корректно делать строгое сравнение, не включая книги, опубликованные 1 января 2000 года, но по опыту обычно нужны данные с самого начала года. В подобных ситуациях стоит уточнять что именно нужно или сделать оба варианта, если это невозможно.
    
💡 Желательно сопровождать запрос комментариями (их можно начинать с <tt>--</tt>) о том, что в нём происходит, это хорошая привычка, она очень облегчает чтение запросов, особенно длинных.
</div>

### Задача №2

Далее пойдут задачи посложнее; выведем информацию и количестве обзоров и среднюю оценку для каждой книги.

In [10]:
# временные таблицы с количеством обзоров и средней оценкой каждой книги
# затем объединяем данные с книжной таблицей по book_id

query = """
        WITH rew AS (SELECT book_id,
                            COUNT(review_id) as count_reviews
                     FROM reviews
                     GROUP BY book_id),
             rat AS (SELECT book_id,
                            AVG(rating) as average_rating
                     FROM ratings
                     GROUP BY book_id)

        SELECT b.title,
               rew.count_reviews,
               rat.average_rating
        FROM books as b
        LEFT OUTER JOIN rew ON b.book_id=rew.book_id
        LEFT OUTER JOIN rat ON b.book_id=rat.book_id
        ORDER BY average_rating DESC;
        """

pd.io.sql.read_sql(query, con = engine) 

title  count_reviews  \
0                        A Dirty Job (Grim Reaper  #1)            4.0   
1                             Tai-Pan (Asian Saga  #2)            2.0   
2              School's Out—Forever (Maximum Ride  #2)            3.0   
3                 Piercing the Darkness (Darkness  #2)            2.0   
4    The War of Art: Break Through the Blocks & Win...            2.0   
..                                                 ...            ...   
995  The World Is Flat: A Brief History of the Twen...            3.0   
996                  His Excellency: George Washington            2.0   
997                                              Junky            2.0   
998                                      Drowning Ruth            3.0   
999                               Harvesting the Heart            2.0   

     average_rating  
0              5.00  
1              5.00  
2              5.00  
3              5.00  
4              5.00  
..              ...  
995            2.25  
996            2.00  
997            2.00  
998            2.00  
999            1.50  

[1000 rows x 3 columns]

<div class="alert alert-success"><b>Комментарий ревьюера</b>&nbsp;👍<br>  
Очень хорошо, что книги сгруппированы книги по id, а не названию и получилось отдельно вывести значения по всем. Также правильно, что есть сортировка.
</div>

Итого в таблице оказались все книги (а их 1000 штук), количество обзоров и средняя оценка, этот запрос отсортирован по рейтингу, так видно, что худшая книга получила среднюю оценку 1,5 на основе двух обзоров, а лучшая A dirty Job = максимальный балл после четырех рецензий. Не то чтобы много рецензий для такой высокой оценки..

### Задача №3

Следующим запросом посмотрим издательства, которые выпустили больше всего книг, по самим книгами будет ограничение от 50 страниц и больше.

In [11]:
# временная таблица с книгами толщиной 50 страниц и больше,
# объединяем с данными об издательствах и сортируем по убыванию

query = """
        WITH b AS (SELECT publisher_id,
                          title
                   FROM books
                   WHERE num_pages >= 50)

        SELECT p.publisher,
               COUNT(b.title) as amount_books
        FROM publishers as p
        INNER JOIN b ON p.publisher_id=b.publisher_id
        GROUP BY p.publisher
        ORDER BY amount_books DESC;
        """

pd.io.sql.read_sql(query, con = engine) 

publisher  amount_books
0                                        Penguin Books            42
1                                              Vintage            31
2                             Grand Central Publishing            25
3                                     Penguin Classics            24
4                                     Ballantine Books            19
..                                                 ...           ...
329                                         Turtleback             1
330  Atheneum Books for Young Readers: Richard Jack...             1
331                                     Penguin Signet             1
332                                    Victor Gollancz             1
333                      Harvard Business Review Press             1

[334 rows x 2 columns]

Из 340 издателей только 6 не вошли в сводную таблицу, остальные 334 активно выпускают книги толще 50 страниц. Лидер среди них - издательство **Penguin Books**, выпущено 42 книги.

<div class="alert alert-warning"><b>Комментарий ревьюера</b>&nbsp;⚠️<br>
Всё верно, но нас просили вывести только одно издательство.
</div>


### Задача №4

Определим автора с самой высокой средней оценкой книг, причем книги должны иметь 50 и более оценок.

In [12]:
# временная таблица идентификаторов книг с 50 и более оценками
# затем объединяем информацию с авторами через таблицу books и считаем среднее

query = """
        WITH rat AS (SELECT book_id,
                            AVG(rating) as average_rating,
                            COUNT(rating) as amount_ratings
                     FROM ratings
                     GROUP BY book_id
                     HAVING COUNT(rating) >= 50)

        SELECT a.author,
               AVG(rat.average_rating) as average_rating
        FROM rat
        LEFT OUTER JOIN books AS b ON b.book_id=rat.book_id
        LEFT OUTER JOIN authors AS a ON a.author_id=b.author_id
        GROUP BY a.author
        ORDER BY average_rating DESC;
        """

pd.io.sql.read_sql(query, con = engine) 

author  average_rating
0                          J.K. Rowling/Mary GrandPré        4.283844
1                   Markus Zusak/Cao Xuân Việt Khương        4.264151
2                                      J.R.R. Tolkien        4.258446
3                                   Louisa May Alcott        4.192308
4                                        Rick Riordan        4.080645
5                                     William Golding        3.901408
6                                       J.D. Salinger        3.825581
7            Paulo Coelho/Alan R. Clarke/Özdemir İnce        3.789474
8   William Shakespeare/Paul Werstine/Barbara A. M...        3.787879
9                                           Dan Brown        3.754540
10                                         Lois Lowry        3.750000
11            George Orwell/Boris Grabnar/Peter Škerl        3.729730
12                                    Stephenie Meyer        3.662500
13                                     John Steinbeck        3.622951

Авторов с оценками книг более 50 нашлось всего восемнадцать, автором с самой высокой средней оценкой книг оказалась **Дж.К.Роулинг (средний рейтинг 4.2838)**. Первая тройка очень близка и отличается лишь сотыми долями оценок.

<div class="alert alert-success"><b>Комментарий ревьюера</b>&nbsp;👍<br>
Сложное задание сделано с первой попытки! Круто! Хорошо что смотришь на все данные, но нас просили тут также вывести одного автора.
</div>

### Задача №5

Наверняка в базе есть пользователи, которые активно пишут рецензии и ставят оценки прочитанным книгам. Если к тому же отсортировать пользователей, которые поставили больше 50 оценок, мы можем вывести среднее количество обзоров от самых активных читателей.

In [13]:
# делаем временную таблицу количества обзоров, в которой подзапросом фильтруем пользователей с 50 и более оценками
# затем выводим средний показатель

query = """  
        WITH a AS (SELECT r.username,
                          COUNT(r.book_id) as amount_reviews
                   FROM reviews as r 
                   WHERE r.username IN (SELECT username    
                                        FROM ratings
                                        GROUP BY username
                                        HAVING COUNT(rating) > 50)
                                        GROUP BY r.username)
               
        SELECT ROUND(AVG(a.amount_reviews)) as average_reviews
        FROM a;
        """

pd.io.sql.read_sql(query, con = engine) 

average_reviews
0             24.0

В среднем на пользователя, оставившего больше 50 оценок, приходится 24 обзора прочитанных книг. 

<div class="alert alert-success"><b>Комментарий ревьюера</b>&nbsp;✔️<br>
И это... правильный ответ!
</div>

## Выводы

Итак, мы подключились к базе данных и произвели несколько SQL-запросов, вот какие результаты получились:

1. В таблице books представлено 1000 книг, из них большая часть (819 штук) выпущено после 01 января 2000 года.


2. Средний рейтинг всех книг находится в диапазоне от 1,5 до 5 баллов, лучшая книга **A dirty Job**, однако на нее опубликовано всего четыре рецензии, худшая = **Harvesting the Heart** с двумя рецензиями. Чаще всего, чем больше обзоров и оценок, тем больше усредняется итоговый книжный балл. 


3. В базе данных представлено 340 издателей, из них 6 занимаются брошюрами и книгами тоньше 50 страниц; среди остальных лидером по изданию книг является предприятие **Penguin Books**, у них опубликовано 42 книги.


4. 18 авторов попали в список самых популярных писателей по количеству оценок на их книги (50 и более). Лидером средней оценки по своим книгам стала **Джоан Роулинг**, рейтинг составил в среднем 4.2838 баллов.


5. Лишь небольшое число пользователей активно публикует обзоры и ставит оценки; в среднем на читателя, оставившего более 50 оценок, приходится **24 опубликованных рецензии** на прочитанные книги.

<div class="alert alert-success"><b>Комментарий ревьюера</b>&nbsp;✔️<br>
Круто, что есть выводы!
</div>

![Апплодисменты](https://i.gifer.com/DhG5.gif "Хлоп-хлоп!")